sdks/java/harness/src/test/java/org/apache/beam/fn/harness/state/OrderedListUserStateTest.java

```diff
@@ -0,0 +1,597 @@
+                Collections.singletonList(A4),
+                createOrderedListStateKey("A", 2),
+                asList(A2, B2),
+                createOrderedListStateKey("A", 3),
+                Collections.singletonList(A3)));
+
+    OrderedListUserState<String> userState =
+        new OrderedListUserState<>(
+            Caches.noop(),
+            fakeClient,
+            "instructionId",
+            createOrderedListStateKey("A"),
+            StringUtf8Coder.of());
+
+    Iterable<TimestampedValue<String>> stateBeforeClear = userState.read();
+    userState.clear();
+    assertArrayEquals(
+        asList(A1, B1, A2, B2, A3, A4).toArray(),
+        Iterables.toArray(stateBeforeClear, TimestampedValue.class));
+    assertThat(userState.read(), is(emptyIterable()));
+
+    userState.asyncClose();
+    assertThrows(IllegalStateException.class, () -> userState.clear());
+  }
+
+  @Test
+  public void testAddAndClearRange() throws Exception {
+    FakeBeamFnStateClient fakeClient =
+        new FakeBeamFnStateClient(
+            TimestampedValueCoder.of(StringUtf8Coder.of()),
+            ImmutableMap.of(
+                createOrderedListStateKey("A", 1),
+                Collections.singletonList(A1),
+                createOrderedListStateKey("A", 3),
+                Collections.singletonList(A3),
+                createOrderedListStateKey("A", 4),
+                Collections.singletonList(A4)));
+
+    OrderedListUserState<String> userState =
+        new OrderedListUserState<>(
+            Caches.noop(),
+            fakeClient,
+            "instructionId",
+            createOrderedListStateKey("A"),
+            StringUtf8Coder.of());
+
+    // add to a non-existing timestamp, clear, and then add
+    userState.add(A2);
+    Iterable<TimestampedValue<String>> stateBeforeFirstClearRange = userState.read();
+    userState.clearRange(Instant.ofEpochMilli(2), Instant.ofEpochMilli(3));
+    assertArrayEquals(
+        asList(A1, A2, A3, A4).toArray(),
+        Iterables.toArray(stateBeforeFirstClearRange, TimestampedValue.class));
+    assertArrayEquals(
+        asList(A1, A3, A4).toArray(), Iterables.toArray(userState.read(), TimestampedValue.class));
+    userState.add(B2);
+    assertArrayEquals(
+        asList(A1, A2, A3, A4).toArray(),
+        Iterables.toArray(stateBeforeFirstClearRange, TimestampedValue.class));
+    assertArrayEquals(
+        asList(A1, B2, A3, A4).toArray(),
+        Iterables.toArray(userState.read(), TimestampedValue.class));
+
+    // add to an existing timestamp, clear, and then add
+    userState.add(B1);
+    userState.clearRange(Instant.ofEpochMilli(1), Instant.ofEpochMilli(2));
+    assertArrayEquals(
+        asList(A1, A2, A3, A4).toArray(),
+        Iterables.toArray(stateBeforeFirstClearRange, TimestampedValue.class));
+    assertArrayEquals(
+        asList(B2, A3, A4).toArray(), Iterables.toArray(userState.read(), TimestampedValue.class));
+    userState.add(B1);
```

Comment:
It would be more interesting if this test added a new value (C1) so we can verify we aren't just rereading the cleared value from line 332.

Reply:
The scenario you referred to has been covered by the test above, where A2 is cleared and then B2 is added.

https://github.com/apache/beam/blob/90b7f7d8cdcb1b9bc871ce6b62498a65e9bb0e90/sdks/java/harness/src/test/java/org/apache/beam/fn/harness/state/OrderedListUserStateTest.java#L315-L323